kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaTopicManager.java

```diff
@@ -110,33 +109,32 @@ public void updateCtx() {
     // topicName is in pulsar format. e.g. persistent://public/default/topic-partition-0
     // future will complete with null when topic not owned by this broker, or meet error.
     public CompletableFuture<KafkaTopicConsumerManager> getTopicConsumerManager(String topicName) {
-        return consumerTopicManagers.computeIfAbsent(
-            topicName,
-            t -> {
-                CompletableFuture<PersistentTopic> topic = getTopic(t);
-                checkState(topic != null);
-
-                return topic.thenApply(t2 -> {
-                    if (log.isDebugEnabled()) {
-                        log.debug("[{}] Call getTopicConsumerManager for {}, and create TCM for {}.",
-                            requestHandler.ctx.channel(), topicName, t2);
-                    }
-
-                    if (t2 == null) {
-                        return null;
-                    }
-                    // return consumer manager
-                    return new KafkaTopicConsumerManager(requestHandler, t2);
-                });
+        CompletableFuture<KafkaTopicConsumerManager> consumerManagerFuture = new CompletableFuture<>();
+        if (consumerTopicManagers.containsKey(topicName)) {
+            return consumerTopicManagers.get(topicName);
+        } else {
+            CompletableFuture<PersistentTopic> topic = getTopic(topicName);
+            if (topic != null) {
+                consumerManagerFuture.complete(new KafkaTopicConsumerManager(requestHandler, topic.join()));
+                consumerTopicManagers.put(topicName, consumerManagerFuture);
+            } else {
+                consumerManagerFuture.complete(null);
             }
-        );
+            return consumerManagerFuture;
+        }
     }
```

Comment:
Could it be changed to `computeIfAbsent`? Because `ConcurrentHashMap#computeIfAbsent` is an atomic operation where combination of `containsKey` and `get`/`put` is not.

Reply:
Makes sense. But these changes make `getTopic` and `getTopicConsumerManager` not thread-safe anymore. Should we just use a normal `HashMap` and `synchronized` to make these methods thread safe? Because there're somewhere else that accesses the `topics`.

Reply:
Also I found that after this change, use `CompletableFuture<KafkaTopicConsumerManager>` as a return value is meaningless because `consumerTopicManagers.put(...)` is only invoked in `getTopicConsumerManager` and the future to add is always a completed future.